File: GoonWorld/GoonEngine/components/AnimationComponent.cs
using GoonEngine.Content;
using GoonEngine.Interfaces;
using GoonEngine.Models;

namespace GoonEngine.Components;
public class AnimationComponent<T> : Component, IDraw where T : GameObject
{
    public string CurrentAnimation;
    private Animator<T> _animator;
    private Animation<T> _currentAnimation;
    private AsepriteDocument _currentAnimationDocument;
    public float _animationSpeed = 1.0f;
    private double _secondsThisFrame = 0;
    private int currentFrame = 0;
    public Rect SpriteImageRect;
    public bool Mirror { get; set; } = false;
    public Point Offset;
    public byte SizeMultiplier = 1;
    public Rect DrawImageRect => new Rect(Parent.Location.X + Offset.X, Parent.Location.Y + Offset.Y, SpriteImageRect.Width * SizeMultiplier, SpriteImageRect.Height * SizeMultiplier);
    public bool Visible { get; set; } = true;


    public AnimationComponent(Animator<T> animator)
    {
        CurrentAnimation = animator.DefaultAnimation;
        _currentAnimation = animator.Animations[CurrentAnimation];
        _animator = animator;
        _currentAnimationDocument = _animator.BaseDocument;
    }

    public override void Update()
    {
        // Check to see if we should change to a different animation.
        if (_animator.CheckAnimationState(this) is AnimatorTransitionArgs<T> thing)
        {
            ChangeAnimation(thing);
        }
        _secondsThisFrame += GameObject.DeltaTime.TotalMilliseconds * _animationSpeed;
        // Check to see if we should change to the next frame
        var frameSeconds = _currentAnimationDocument.frames[currentFrame].duration;
        if (_secondsThisFrame >= _currentAnimationDocument.frames[currentFrame].duration)
        {
            _secondsThisFrame -= frameSeconds;
            currentFrame = _currentAnimation.Looping ? (currentFrame + 1 <= _currentAnimation.EndFrame) ? currentFrame + 1 : _currentAnimation.StartFrame : currentFrame;
            SpriteImageRect = new Rect(
                _currentAnimationDocument.frames[currentFrame].frame.x,
                _currentAnimationDocument.frames[currentFrame].frame.y,
                _currentAnimationDocument.frames[currentFrame].frame.w,
                _currentAnimationDocument.frames[currentFrame].frame.h
            );
        }
    }

    private void ChangeAnimation(AnimatorTransitionArgs<T> args)
    {
        // Debug.InfoMessage($"I should be switching to {args.Animation.Name}");
        CurrentAnimation = args.Animation.Name;
        _currentAnimation = args.Animation;
        _currentAnimationDocument = args.Document;
        currentFrame = _currentAnimation.StartFrame;
        SpriteImageRect = new Rect(
            _currentAnimationDocument.frames[currentFrame].frame.x,
            _currentAnimationDocument.frames[currentFrame].frame.y,
            _currentAnimationDocument.frames[currentFrame].frame.w,
            _currentAnimationDocument.frames[currentFrame].frame.h
        );
    }

    public void Draw()
    {
        if (!Enabled || !Visible || _currentAnimation == null)
            return;
        var drawRect = DrawImageRect;
        Api.Rendering.DrawTexture(_currentAnimation.Image.LoadedImagePtr, ref SpriteImageRect, ref drawRect);
    }
    public override void OnComponentAdd(GameObject parent)
    {
        base.OnComponentAdd(parent);
        GameObject.CurrentDrawComponents.Add(this);
    }

}

public class Animation<T>
{
    public string Name;
    public bool Looping;
    public Image? Image;
    public int StartFrame;
    public int EndFrame;
    public List<AnimationTransition<T>> Transitions = new();
    public Animation(string name, bool looping = true)
    {
        Name = name;
        Looping = looping;
    }
}

public class AnimationTransition<T>
{
    public string TransitionAnimationTag;
    public Func<T, bool> TransitionCondition;

    public AnimationTransition(string transitionAnimationTag, Func<T, bool> transitionCondition)
    {
        TransitionAnimationTag = transitionAnimationTag;
        TransitionCondition = transitionCondition;
    }
}

File: src/GoonWorld/components/AnimationComponent.cpp
#include <GoonWorld/gnpch.hpp>
#include <GoonWorld/core/Game.hpp>
#include <GoonWorld/components/AnimationComponent.hpp>
#include <GoonWorld/animation/Animator.hpp>
#include <GoonWorld/animation/AnimationTransition.hpp>
#include <GoonWorld/common/TimeSpan.hpp>
#include <GoonWorld/base/GameObject.hpp>
#include <GoonWorld/core/Camera.hpp>
// #include <GoonEngine/SdlSurface.h>

// temp
// #include <GoonEngine/Shader.h>
// #include <GoonEngine/Sprite.h>
#include <GoonEngine/color.h>
using namespace GoonWorld;

// extern geShader *shader;

AnimationComponent::AnimationComponent(std::string animator)
    : Component((int)ComponentTypes::Animation), AnimationSpeed(1.0), _animator(Animator::GetAnimator(animator)), _visible(true)
{
    _currentAnimationDocument = &_animator->_loadedDocument;
    ChangeAnimation(_animator->_defaultAnimation);
    // _sprite = geSpriteRendererNew(shader);
}

AnimationComponent::AnimationComponent(std::string animator, Point offset)
    : AnimationComponent(animator)
{
    _visible = true;
    _offset = offset;
}
void AnimationComponent::Update()
{
    for (auto &transition : _transitions)
    {
        if (transition->CurrentAnimation != _currentAnimationString)
            continue;
        if (*transition->Condition == transition->ConditionMatch)
        {
            ChangeAnimation(transition->NextAnimation);
            break;
        }
    }
    if (_currentAnimation->StartFrame == _currentAnimation->EndFrame)
        return;
    _secondsThisFrame += GameObject::DeltaTime.GetTotalMilliseconds() * AnimationSpeed;
    auto frameSeconds = _currentAnimationDocument->frames[_currentFrame].duration;
    if (_secondsThisFrame >= frameSeconds)
    {
        _secondsThisFrame -= frameSeconds;
        _currentFrame = _currentAnimation->Looping ? (_currentFrame + 1 <= _currentAnimation->EndFrame) ? _currentFrame + 1 : _currentAnimation->StartFrame : _currentFrame;
        SpriteImageRect = geRectangle{
            _currentAnimationDocument->frames[_currentFrame].frame.x,
            _currentAnimationDocument->frames[_currentFrame].frame.y,
            _currentAnimationDocument->frames[_currentFrame].frame.w,
            _currentAnimationDocument->frames[_currentFrame].frame.h,
        };
    }
}
void AnimationComponent::ChangeAnimation(std::string &anim, int frameInAnim)
{
    auto newAnimation = _animator->GetAnimation(anim);
    if (!newAnimation)
        return;
    _currentAnimation = newAnimation;
    _currentAnimationString = anim;
    _currentFrame = frameInAnim;
    SpriteImageRect = geRectangle{
        _currentAnimationDocument->frames[_currentFrame].frame.x,
        _currentAnimationDocument->frames[_currentFrame].frame.y,
        _currentAnimationDocument->frames[_currentFrame].frame.w,
        _currentAnimationDocument->frames[_currentFrame].frame.h,
    };
}
void AnimationComponent::ChangeAnimation(std::string &nextAnim)
{
    auto newAnimation = _animator->GetAnimation(nextAnim);
    if (!newAnimation)
        return;
    _currentAnimationString = nextAnim;
    _currentAnimation = newAnimation;
    _currentFrame = _currentAnimation->StartFrame;
    _secondsThisFrame = 0;
    SpriteImageRect = geRectangle{
        _currentAnimationDocument->frames[_currentFrame].frame.x,
        _currentAnimationDocument->frames[_currentFrame].frame.y,
        _currentAnimationDocument->frames[_currentFrame].frame.w,
        _currentAnimationDocument->frames[_currentFrame].frame.h,
    };
}
void AnimationComponent::Draw()
{
    if (!IsEnabled() || _currentAnimation == nullptr)
        return;
    auto rect = GetDrawRect();
    // geDrawTextureWithCameraOffset(_currentAnimation->Image, &SpriteImageRect, &rect, Mirror);
    GetGame().SpritebatchDraw(_currentAnimation->Image->GetTexture(), rect,
                              0, geColor{1, 1, 1, 1}, SpriteImageRect, Mirror);
    // GetGame().SpritebatchDraw(_currentAnimation->Image->GetTexture(), vec2{(float)rect.x, (float)rect.y},
    //                           vec2{(float)rect.w, (float)rect.h}, 0, vec4{1, 1, 1, 1}, vec2{(float)SpriteImageRect.x, (float)SpriteImageRect.y},
    //                           vec2{(float)SpriteImageRect.w, (float)SpriteImageRect.h}, Mirror);
    // geSpriteRendererDraw(_sprite, _currentAnimation->Image->GetTexture(), vec2{(float)rect.x, (float)rect.y},
    //                      vec2{(float)rect.w, (float)rect.h}, 0, vec4{1, 1, 1, 1}, vec2{(float)SpriteImageRect.x, (float)SpriteImageRect.y},
    //                      vec2{(float)SpriteImageRect.w, (float)SpriteImageRect.h}, Mirror);
}

void AnimationComponent::OnComponentAdd(GameObject &parent)
{
    Component::OnComponentAdd(parent);
    GetGame().AddDrawObject(this);
}

geRectangle AnimationComponent::GetDrawRect()
{
    return geRectangle{Parent()->Location().x + _offset.x, Parent()->Location().y + _offset.y, SpriteImageRect.w * SizeMultiplier, SpriteImageRect.h * SizeMultiplier};
}

void AnimationComponent::AddTransition(std::string from, std::string to, bool matchCondition, bool *matchRef)
{
    _transitions.push_back(std::make_unique<AnimationTransition>(from, to, matchCondition, matchRef));
}

void AnimationComponent::ChangeLayer(int layer)
{
    GetGame().ChangeDrawObjectLayer(this, layer);
    _layer = layer;
}